File: agenkit-scala/src/main/scala/io/agenkit/observability/TracingAgent.scala
package io.agenkit.observability

import io.agenkit.core.*
import org.slf4j.LoggerFactory
import scala.concurrent.{ExecutionContext, Future}
import java.util.UUID
import java.util.concurrent.atomic.AtomicLong

case class TraceSpan(
  traceId: String,
  spanId: String,
  operation: String,
  startMs: Long,
  endMs: Long,
  success: Boolean
)

/** Wraps an agent with distributed tracing spans, propagating trace IDs via message metadata. */
class TracingAgent(
  inner: Agent,
  serviceName: String = "agenkit"
) extends Agent:
  private val logger          = LoggerFactory.getLogger(getClass)
  private val _processedCount = new AtomicLong(0)
  private val spans           = collection.mutable.ListBuffer[TraceSpan]()

  def name: String               = s"tracing(${inner.name})"
  def capabilities: List[String] = inner.capabilities :+ "tracing"
  def getSpans: List[TraceSpan]  = spans.toList

  def process(message: Message)(using ExecutionContext): Future[Message] =
    _processedCount.incrementAndGet()
    val traceId = message.metadata.get("trace_id").map(_.toString).getOrElse(UUID.randomUUID().toString)
    val spanId  = UUID.randomUUID().toString
    val start   = System.currentTimeMillis()
    logger.debug(s"[$serviceName] trace=$traceId span=$spanId START ${inner.name}")
    inner.process(message.copy(metadata = message.metadata + ("trace_id" -> traceId)))
      .map { response =>
        val end = System.currentTimeMillis()
        spans += TraceSpan(traceId, spanId, inner.name, start, end, success = true)
        logger.debug(s"[$serviceName] trace=$traceId span=$spanId END latency=${end - start}ms")
        response
      }
      .recoverWith { case ex =>
        val end = System.currentTimeMillis()
        spans += TraceSpan(traceId, spanId, inner.name, start, end, success = false)
        logger.warn(s"[$serviceName] trace=$traceId span=$spanId ERROR ${ex.getMessage}")
        Future.failed(ex)
      }

  def introspect(): IntrospectionResult =
    IntrospectionResult(
      name = name,
      metadata = Map("spanCount" -> spans.size, "service" -> serviceName),
      capabilities = capabilities,
      processedMessages = _processedCount.get()
    )

File: agenkit-java/src/main/java/io/agenkit/observability/TracingAgent.java
package io.agenkit.observability;

import io.agenkit.core.Agent;
import io.agenkit.core.IntrospectionResult;
import io.agenkit.core.Message;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

/**
 * Agent wrapper that adds distributed tracing spans.
 */
public final class TracingAgent implements Agent {

    private static final Logger log = LoggerFactory.getLogger(TracingAgent.class);

    private final Agent inner;
    private final List<Span> spans = new ArrayList<>();

    public TracingAgent(Agent inner) {
        this.inner = inner;
    }

    @Override
    public String getName() { return inner.getName() + "[tracing]"; }

    @Override
    public List<String> getCapabilities() { return inner.getCapabilities(); }

    @Override
    public CompletableFuture<Message> process(Message message) {
        String traceId = (String) message.getMetadata().getOrDefault(
                "trace_id", UUID.randomUUID().toString());
        String spanId = UUID.randomUUID().toString().substring(0, 8);
        Instant start = Instant.now();

        log.debug("trace={} span={} agent={} start", traceId, spanId, inner.getName());

        return inner.process(message.withMetadata("trace_id", traceId))
                .thenApply(response -> {
                    long durationMs = Instant.now().toEpochMilli() - start.toEpochMilli();
                    Span span = new Span(traceId, spanId, inner.getName(), start,
                            durationMs, null);
                    synchronized (spans) { spans.add(span); }
                    log.debug("trace={} span={} agent={} duration={}ms",
                            traceId, spanId, inner.getName(), durationMs);
                    return response
                            .withMetadata("trace_id", traceId)
                            .withMetadata("span_id", spanId)
                            .withMetadata("duration_ms", durationMs);
                })
                .exceptionallyCompose(ex -> {
                    long durationMs = Instant.now().toEpochMilli() - start.toEpochMilli();
                    Span span = new Span(traceId, spanId, inner.getName(), start,
                            durationMs, ex.getMessage());
                    synchronized (spans) { spans.add(span); }
                    log.warn("trace={} span={} agent={} error={}",
                            traceId, spanId, inner.getName(), ex.getMessage());
                    return CompletableFuture.failedFuture(ex);
                });
    }

    public synchronized List<Span> getSpans() { return List.copyOf(spans); }

    @Override
    public IntrospectionResult introspect() { return inner.introspect(); }

    public record Span(
            String traceId,
            String spanId,
            String agentName,
            Instant startTime,
            long durationMs,
            String error) {}
}
